dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Filter/BasicFilterTestsBase.cs

```diff
@@ -0,0 +1,283 @@
+        => this.TestFilterAsync(r => !(r.Int == 8 || r.String == "foo"));
+
+    [ConditionalFact]
+    public virtual Task Not_over_bool()
+        => this.TestFilterAsync(r => !r.Bool);
+
+    #endregion Logical operators
+
+    #region Contains
+
+    [ConditionalFact]
+    public virtual Task Contains_over_field_string_array()
+        => this.TestFilterAsync(r => r.StringArray.Contains("x"));
+
+    [ConditionalFact]
+    public virtual Task Contains_over_field_string_List()
+        => this.TestFilterAsync(r => r.StringList.Contains("x"));
+
+    [ConditionalFact]
+    public virtual Task Contains_over_inline_int_array()
+        => this.TestFilterAsync(r => new[] { 8, 10 }.Contains(r.Int));
+
+    [ConditionalFact]
+    public virtual Task Contains_over_inline_string_array()
+        => this.TestFilterAsync(r => new[] { "foo", "baz", "unknown" }.Contains(r.String));
+
+    [ConditionalFact]
+    public virtual Task Contains_over_inline_string_array_with_weird_chars()
+        => this.TestFilterAsync(r => new[] { "foo", "baz", "un  , ' \"" }.Contains(r.String));
+
+    [ConditionalFact]
+    public virtual Task Contains_over_captured_string_array()
+    {
+        var array = new[] { "foo", "baz", "unknown" };
+
+        return this.TestFilterAsync(r => array.Contains(r.String));
+    }
+
+    #endregion Contains
+
+    [ConditionalFact]
+    public virtual Task Captured_variable()
+    {
+        // ReSharper disable once ConvertToConstant.Local
+        var i = 8;
+
+        return this.TestFilterAsync(r => r.Int == i);
+    }
+
+    #region Legacy filter support
+
+    [ConditionalFact]
+    [Obsolete("Legacy filter support")]
+    public virtual Task Legacy_equality()
+        => this.TestLegacyFilterAsync(
+            new VectorSearchFilter().EqualTo("Int", 8),
+            r => r.Int == 8);
+
+    [ConditionalFact]
+    [Obsolete("Legacy filter support")]
+    public virtual Task Legacy_And()
+        => this.TestLegacyFilterAsync(
+            new VectorSearchFilter().EqualTo("Int", 8).EqualTo("String", "foo"),
+            r => r.Int == 8);
+
+    [ConditionalFact]
+    [Obsolete("Legacy filter support")]
+    public virtual Task Legacy_AnyTagEqualTo_array()
+        => this.TestLegacyFilterAsync(
+            new VectorSearchFilter().AnyTagEqualTo("StringArray", "x"),
+            r => r.StringArray.Contains("x"));
+
+    [ConditionalFact]
+    [Obsolete("Legacy filter support")]
+    public virtual Task Legacy_AnyTagEqualTo_List()
+        => this.TestLegacyFilterAsync(
+            new VectorSearchFilter().AnyTagEqualTo("StringList", "x"),
+            r => r.StringArray.Contains("x"));
+
+    #endregion Legacy filter support
+
+    protected virtual async Task TestFilterAsync(
+        Expression<Func<FilterRecord<TKey>, bool>> filter,
+        bool expectZeroResults = false,
+        bool expectAllResults = false)
+    {
+        var expected = fixture.TestData.AsQueryable().Where(filter).OrderBy(r => r.Key).ToList();
```

Comment:
This is very clever (the tests don't need to specify the expected results which makes them very simple and we are comparing the results against very reliable LINQ implementation)! 🥇

Reply:
Well I can't really take the credit for it - most of EF's query tests are done exactly like this ;) Yeah, using LINQ as a querying abstraction is really nice, since you can use the IEnumerable implementation as a free, authoritative oracle.